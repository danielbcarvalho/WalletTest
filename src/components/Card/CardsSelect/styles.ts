import { FlatList } from 'react-native';
import { css } from 'styled-components';
import styled from 'styled-components/native';

interface Props {
  cardTop: boolean;
}

export const Container = styled.View`
  flex: 1;
  width: 100%;
`;

export const UseButton = styled.TouchableOpacity`
  width: 100%;
  align-items: center;
`;

export const CardFlatList = styled.FlatList`` as typeof FlatList;

export const Text = styled.Text`
  ${({ theme }) => css`
    font-size: ${theme.typography.p.fontSize};
    line-height: ${theme.typography.p.lineHeight};
    color: ${theme.pallete.textOnBackground.default};
  `}
  margin-top: 15px;
`;

export const CardContainer = styled.TouchableOpacity<Props>`
  margin-bottom: ${({ cardTop }) => (cardTop ? '0' : '-105px')};
`;
