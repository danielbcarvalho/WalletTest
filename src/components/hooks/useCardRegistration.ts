import { Alert } from 'react-native';
import uuid from 'react-native-uuid';
import { useNavigation } from '@react-navigation/native';
import { useMutation, useQueryClient } from '@tanstack/react-query';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';

import { ParamsList } from '../../Routes';
import { cardService } from '../../services/card';
import { QueryKeys } from '../../services/QueryKeys';

type CardRegistrationScreenProps = NativeStackNavigationProp<
  ParamsList,
  'CardRegistration'
>;

interface CardRegistrationFormData {
  cvv: string;
  name: string;
  number: string;
  expiry: string;
}

export function useCardRegistration() {
  const navigation = useNavigation<CardRegistrationScreenProps>();
  const queryClient = useQueryClient();

  const registerCard = useMutation(cardService.register, {
    onSuccess: data => {
      queryClient.invalidateQueries([QueryKeys.CARD_LIST]);
      navigation.push('Home');
      navigation.navigate('CardRegistrationSuccess', { card: data });
    },
    onError: (error: any) => {
      Alert.alert('Error', error.message);
    },
  });

  const handleCardRegisterForm = async (data: CardRegistrationFormData) => {
    await registerCard.mutateAsync({
      id: uuid.v4() as string,
      kind: 'black',
      ...data,
    });
  };

  return handleCardRegisterForm;
}
