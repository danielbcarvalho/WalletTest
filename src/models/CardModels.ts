export interface Card {
  id: string;
  cvv: string;
  name: string;
  expiry: string;
  number: string;
  kind: 'black' | 'green';
}
