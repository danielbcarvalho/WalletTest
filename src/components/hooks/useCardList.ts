import { Alert } from 'react-native';
import { useEffect, useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import { useQuery, useQueryClient } from '@tanstack/react-query';

import { Card } from '../../models/CardModels';
import { cardService } from '../../services/card';
import { QueryKeys } from '../../services/QueryKeys';
import { isCardOnTheTop, topCardIndex } from '../../utils/card';
import { CardListScreenProps } from '../../screens/CardListScreen';

const FIVE_MINUTES = 1000 * 60 * 5;

export function useCardList() {
  const queryClient = useQueryClient();
  const { navigate } = useNavigation<CardListScreenProps>();
  const [useThisCard, setUseThisCard] = useState<boolean>(false);
  const [payWithThisCard, setPayWithThisCard] = useState<boolean>(false);

  const {
    isError,
    data: cards,
    isInitialLoading,
  } = useQuery<Card[]>([QueryKeys.CARD_LIST], () => cardService.list(), {
    staleTime: FIVE_MINUTES,
  });

  function setCardOnTop(selectedCardId: string) {
    if (cards) {
      const cardSelectedIndex = cards.findIndex(
        card => card.id === selectedCardId,
      );

      if (!isCardOnTheTop(cardSelectedIndex, cards)) {
        queryClient.setQueryData(
          [QueryKeys.CARD_LIST],
          [cards[topCardIndex(cards)], cards[cardSelectedIndex]],
        );
      }
    }
  }

  function onPayWithThisCard(value: boolean) {
    setPayWithThisCard(value);
  }

  function onUseThisCard() {
    setUseThisCard(true);
    navigate('CardPayment');
  }

  useEffect(() => {
    if (isError) {
      Alert.alert('Something went wrong');
      navigate('Home');
    }

    if (isInitialLoading) {
      navigate('WalletAnimatedScreen');
    }
  }, [cards, isError, navigate, isInitialLoading]);

  return {
    cards: cards ?? [],
    isError,
    useThisCard,
    setCardOnTop,
    onUseThisCard,
    payWithThisCard,
    onPayWithThisCard,
    isLoading: isInitialLoading,
  };
}
