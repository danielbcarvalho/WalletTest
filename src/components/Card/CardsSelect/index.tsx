import React from 'react';
import { useTranslation } from 'react-i18next';

import Card from '..';
import { Card as CardModel } from '../../../models/CardModels';

import {
  Text,
  Container,
  UseButton,
  CardFlatList,
  CardContainer,
} from './styles';
import { useCardList } from '../../hooks/useCardList';
import { isCardOnTheTop } from '../../../utils/card';

function CardsSelect() {
  const { t } = useTranslation();

  const { cards, setCardOnTop, onUseThisCard } = useCardList();

  if (cards.length === 0) {
    return <Text>{t('ops, no card registered')}</Text>;
  }

  function renderUseCardButton() {
    return (
      <UseButton onPress={onUseThisCard}>
        <Text>{t('use this card')}</Text>
      </UseButton>
    );
  }

  function renderCard({ item, index }: { item: CardModel; index: number }) {
    return (
      <CardContainer
        activeOpacity={0.7}
        onPress={() => setCardOnTop(item.id)}
        cardTop={isCardOnTheTop(index, cards)}
        disabled={isCardOnTheTop(index, cards)}>
        <Card card={item} />
      </CardContainer>
    );
  }

  return (
    <Container>
      <CardFlatList
        data={cards}
        bounces={false}
        renderItem={renderCard}
        ListFooterComponent={renderUseCardButton}
        keyExtractor={item => item.id.toString()}
        contentContainerStyle={{ flexGrow: 1, justifyContent: 'center' }}
      />
    </Container>
  );
}

export default CardsSelect;
