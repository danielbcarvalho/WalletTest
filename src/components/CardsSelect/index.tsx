import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';

import Card from '../Card';
import { Card as CardModel } from '../../models/CardModels';

import {
  Text,
  Container,
  UseButton,
  CardFlatList,
  CardContainer,
} from './styles';

interface CardsSelectProps {
  cards: CardModel[];
  onUseCard: (selectedCard: CardModel) => void;
}

function CardsSelect({ cards, onUseCard }: CardsSelectProps) {
  const { t } = useTranslation();
  const [cardsData, setCardsData] = useState(cards);

  if (cards.length === 0) {
    return <Text>{t('ops, no card registered')}</Text>;
  }

  const topCardIndex = cards.length - 1;

  const isCardOnTheTop = (index: number) => index === topCardIndex;

  function renderUseCardButton() {
    return (
      <UseButton onPress={() => onUseCard(cardsData[topCardIndex])}>
        <Text>{t('use this card')}</Text>
      </UseButton>
    );
  }

  function renderCard({ item, index }: { item: CardModel; index: number }) {
    return (
      <CardContainer
        disabled={isCardOnTheTop(index)}
        activeOpacity={0.7}
        cardTop={isCardOnTheTop(index)}
        onPress={() => setCardOnTheTop(index)}>
        <Card card={item} />
      </CardContainer>
    );
  }

  const setCardOnTheTop = (index: number) => {
    if (!isCardOnTheTop(index)) {
      setCardsData([cardsData[topCardIndex], cardsData[index]]);
    }
  };

  return (
    <Container>
      <CardFlatList
        data={cardsData}
        bounces={false}
        renderItem={renderCard}
        ListFooterComponent={renderUseCardButton}
        keyExtractor={item => item.id.toString()}
      />
    </Container>
  );
}

export default CardsSelect;
