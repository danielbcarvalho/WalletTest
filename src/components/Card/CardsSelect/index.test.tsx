import React from 'react';
import { act } from 'react-test-renderer';
import * as ReactQuery from '@tanstack/react-query';
import { waitFor } from '@testing-library/react-native';
import { renderHook } from '@testing-library/react-hooks';

import { Card } from '../../../models/CardModels';
import {
  render,
  fireEvent,
  createQueryClientWrapper,
} from '../../../../jest/test-utils';

import CardsSelect from '.';
import { useNavigation } from '@react-navigation/native';
import { useCardList } from '../../hooks/useCardList';

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));

const mockCards: Card[] = [
  {
    id: '4ec42ba9-50af-40d2-af90-8312edbd9ca2',
    number: '3529 5435 3355 8720',
    cvv: '017',
    name: 'John Doe',
    expiry: '12/23',
    kind: 'black',
  },
  {
    name: 'Daniel B Carvalho',
    number: '5203 9436 6133 9001',
    cvv: '123',
    expiry: '11/24',
    id: '4ec42ba9-50af-40d2-af90-8312edbd9XXX',
    kind: 'green',
  },
];

const mockSelectedCardIndex = mockCards.length - 1;

describe('CardsSelect', () => {
  const mockNavigate = jest.fn();

  beforeAll(() => {
    useNavigation.mockReturnValue({ navigate: mockNavigate });
  });

  test('renders correctly with cards', () => {
    const component = render(<CardsSelect />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders correctly with no cards', async () => {
    jest.spyOn(ReactQuery, 'useQuery').mockImplementation(
      jest.fn().mockReturnValue({
        data: [],
      }),
    );

    const component = render(<CardsSelect />);

    const { result } = renderHook(() => useCardList(), {
      wrapper: createQueryClientWrapper(),
    });

    await waitFor(() => {
      expect(result.current.cards).toEqual([]);
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('calls onUseCard with correct selected card when "Use Card" button is pressed', async () => {
    jest.spyOn(ReactQuery, 'useQuery').mockImplementation(
      jest.fn().mockReturnValue({
        data: mockCards,
      }),
    );
    const { getByText } = render(<CardsSelect />);

    const { result } = renderHook(() => useCardList(), {
      wrapper: createQueryClientWrapper(),
    });

    await act(async () => {
      fireEvent.press(getByText('use this card'));
    });

    expect(result.current.cards[mockSelectedCardIndex]).toBe(
      mockCards[mockSelectedCardIndex],
    );
  });
});
