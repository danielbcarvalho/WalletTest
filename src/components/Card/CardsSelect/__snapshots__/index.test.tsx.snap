// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsSelect renders correctly with cards 1`] = `
<Text
  style={
    {
      "color": "#FFF",
      "fontSize": 16,
      "lineHeight": 28,
      "marginTop": 15,
    }
  }
>
  ops, no card registered
</Text>
`;

exports[`CardsSelect renders correctly with no cards 1`] = `
<Text
  style={
    {
      "color": "#FFF",
      "fontSize": 16,
      "lineHeight": 28,
      "marginTop": 15,
    }
  }
>
  ops, no card registered
</Text>
`;
